### PR TITLE
[feature] Added support for OpenVPN tls-auth option #174

### LINKED ISSUE #174: [feature] Add support for OpenVPN tls_auth option
Add support for OpenVPN tls_auth option (client or server, automatically detected depending on the mode used) in both the OpenVPN backend and in the OpenWRT backend.

### PR DESCRIPTION
**Pull request by pandafy**
The OpenVPN backend will automatically create a file for the
key present in "tls_auth" field and update the value of the
"tls-auth" parameter.

Closes #174

### CODE REVIEW

File: netjsonconfig/backends/openvpn/schema.py [hunk 1]
```diff
@@ -456,6 +456,16 @@
...
                 },
+                "tls_auth": {
+                    "title": "TLS Auth",
+                    "description": (
+                        "Adds an additional layer of HMAC authentication on top of "
+                        "the TLS control channel to mitigate DoS attacks and "
+                        "attacks on the TLS stack"
+                    ),
+                    "type": "string",
+                    "propertyOrder": 53,
```
Comment (nemesisdesign) [inline netjsonconfig/backends/openvpn/schema.py:467]: Let's add `"format": "textarea",` to display it as a textarea in the UI.

File: netjsonconfig/backends/openvpn/converters.py [hunk 2]
```diff
@@ -46,8 +47,39 @@ def __intermediate_vpn(self, config, remove=[False, 0, '']):
...
         return self.sorted_dict(config)
 
+    def __add_tls_auth_key(self, config):
+        tls_auth = config.get('tls_auth', None)
+        if not tls_auth:
+            return config
+        tls_auth = tls_auth.strip()
+        if len(tls_auth.split(' ')) == 1:
+            # The TLS Auth key is present in the field.
+            ca_path = config.get('ca', '')
```
Comment (nemesisdesign) [inline netjsonconfig/backends/openvpn/converters.py:60]: this code is a bit unclear, are you using the ca path to determine where to put the file? Let's make this clearer.

File: netjsonconfig/backends/openvpn/converters.py [hunk 3]
```diff
@@ -46,8 +47,39 @@ def __intermediate_vpn(self, config, remove=[False, 0, '']):
...
+            tls_auth_path = '/'.join(ca_path.split('/')[:-1] + ['tls_auth.key'])
```
Comment (nemesisdesign) [inline netjsonconfig/backends/openvpn/converters.py:61]: What if there's multiple VPN servers or clients, do the files collide? If yes, please add a test for this and make sure the tls key is unique (eg: each config has a name which can be used for this).

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
Good progress, see below.

File: docs/source/backends/openvpn.rst [hunk 4]
```diff
@@ -171,6 +171,8 @@ Required properties:
 +--------------------------+---------+--------------+-------------------------------------------------------------+
 | ``verb``                 | integer | ``1``        | from ``0`` (disabled) to ``11`` (very verbose)              |
 +--------------------------+---------+--------------+-------------------------------------------------------------+
+| ``tls_auth``             | string  |              | string containing TLS Auth key                              |
```
Comment (nemesisdesign) [inline docs/source/backends/openvpn.rst:174]: let's move this after `tls_cipher`, in general, let's add new options where they make sense, not just at the end.

**Review by nemesisdesign — CHANGES_REQUESTED** [review]

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
I see two issues.

1. the result in OpenWISP Controller doesn't match what I would expect

Screenshot from 2022-03-09 12-55-56

2. I tried creating a server with the TLS auth key, then created the template client, but suprisingly the TLS auth key on the client was not autogenerated. Please add a specific test for the auto client feature.

Please pull my changes: 4145fc129d4a4d52203ae540d98891bcecc3d4a0.

File: netjsonconfig/backends/openvpn/schema.py [hunk 5]
```diff
@@ -190,6 +190,17 @@
...
+                "tls_auth": {
+                    "title": "TLS Auth",
+                    "description": (
+                        "Adds an additional layer of HMAC authentication on top of "
+                        "the TLS control channel to mitigate DoS attacks and "
+                        "attacks on the TLS stack"
+                    ),
+                    "type": "string",
+                    "format": "textarea",
+                    "propertyOrder": 18,
```
Comment (pandafy) [inline netjsonconfig/backends/openvpn/schema.py:202]: It has same `propertyOrder` as `crl_verify`. 

https://github.com/openwisp/netjsonconfig/blob/f2e692cee7d3a26b59c3a8143ddfcbc97816445c/netjsonconfig/backends/openvpn/schema.py#L622-L628

**Comment by pandafy** [comment]
VPN server 
Screenshot from 2022-03-09 22-08-23

VPN Client Template 
Screenshot from 2022-03-09 22-11-38
Screenshot from 2022-03-09 22-11-54

Device
Screenshot from 2022-03-09 22-19-41
Screenshot from 2022-03-09 22-19-52

**Review by nemesisdesign — APPROVED** [review]

**Pull request merged**